core/gracefulpanic/channel.go

```diff
@@ -0,0 +1,16 @@
+package gracefulpanic
+
+import "sync"
+
+var ch = make(chan struct{})
+var panicOnce sync.Once
+
+func Panic() {
+	panicOnce.Do(func() {
+		go close(ch)
+	})
+}
```

Comment:
An important aspect which may be missing is preventing statements after Panic() from being called.  The above seems to continue with execution.  Maybe panic within the panic?

Was there an inspiration for this approach?

Reply:
This is something I've been thinking over.  I can see it from two directions:
- **No panic:** any time the flow of execution reaches this point, it means that the node is already unable to access the DB.  As a result, every such call will naturally error out.  No further work can be done, and no new state can be persisted to the DB.  Meanwhile, the top-down graceful shutdown will eventually reach whatever process is trying to talk to the DB, and will do any remaining cleanup.
- **Panic:** we can panic here, which will unroll this goroutine's stack and call all `defer` statements along the way.  But if this goroutine is unrolling upward while the graceful shutdown is rolling downward, we will need to guard every `.Close` method with a `sync.Once` or we might risk getting into a bad state if something is de-initialized twice in parallel.

Reply:
Re: the inspiration for this approach, I went with a new package containing a package global so that we don't have to thread channels or contexts all throughout the codebase.  I'm generally opposed to globals, but in this case, a global actually fits the semantics of what we want: a universal shutdown signal that takes down the entire process.

Not sure if that answers your question fully.

Reply:
What about an `os.Exit` ?

Reply:
No `defer` statements are executed if you do an `os.Exit`.